File: saral_hr/saral_hr/report/labour_welfare_fund_register/labour_welfare_fund_register.py
# Copyright (c) 2026, sj and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import flt

COMPONENT_NAME = "Employee -Labour Welfare Fund"

MONTH_MAP = {
    'January': 1, 'February': 2, 'March': 3, 'April': 4,
    'May': 5, 'June': 6, 'July': 7, 'August': 8,
    'September': 9, 'October': 10, 'November': 11, 'December': 12
}


def execute(filters=None):
    filters = filters or {}
    validate_filters(filters)
    columns = get_columns()
    data    = get_data(filters)
    return columns, data


def validate_filters(filters):
    if not filters.get("month"):
        frappe.throw("Please select a Month")
    if not filters.get("year"):
        frappe.throw("Please select a Year")


def get_columns():
    return [
        {
            "label": "Sr. No.",
            "fieldname": "sr_no",
            "fieldtype": "Int",
            "width": 70
        },
        {
            "label": "Employee ID",
            "fieldname": "employee_id",
            "fieldtype": "Data",
            "width": 160
        },
        {
            "label": "Employee Name",
            "fieldname": "employee_name",
            "fieldtype": "Data",
            "width": 220
        },
        {
            "label": "Mobile No",
            "fieldname": "mobile_no",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": "Aadhar No",
            "fieldname": "aadhar_no",
            "fieldtype": "Data",
            "width": 180
        },
        {
            "label": "Net Salary",
            "fieldname": "net_salary",
            "fieldtype": "Float",
            "precision": 2,
            "width": 160
        },
        {
            "label": "LWF Amount",
            "fieldname": "lwf_amount",
            "fieldtype": "Float",
            "precision": 2,
            "width": 150
        },
    ]


def get_data(filters):
    month            = filters.get("month")
    year             = filters.get("year")
    employee_filter  = filters.get("employee")   # list from MultiSelectList
    company_filter   = filters.get("company")    # list from MultiSelectList

    month_num  = MONTH_MAP.get(month)
    start_date = f"{year}-{month_num:02d}-01"

    conditions  = ""
    query_params = {"start_date": start_date, "component": COMPONENT_NAME}

    # --- Company filter ---
    if company_filter:
        if isinstance(company_filter, str):
            import json
            company_filter = json.loads(company_filter)
        if company_filter:
            conditions += " AND ss.company IN %(companies)s"
            query_params["companies"] = tuple(company_filter)

    # --- Employee filter ---
    if employee_filter:
        if isinstance(employee_filter, str):
            import json
            employee_filter = json.loads(employee_filter)
        if employee_filter:
            conditions += " AND ss.employee IN %(employees)s"
            query_params["employees"] = tuple(employee_filter)

    rows = frappe.db.sql("""
        SELECT
            ss.employee        AS employee_id,
            ss.employee_name   AS employee_name,
            ss.net_salary      AS net_salary,
            sd.amount          AS lwf_amount
        FROM
            `tabSalary Slip` ss
        INNER JOIN
            `tabSalary Details` sd
            ON  sd.parent           = ss.name
            AND sd.parenttype       = 'Salary Slip'
            AND sd.parentfield      = 'deductions'
            AND sd.salary_component = %(component)s
            AND sd.amount > 0
        WHERE
            ss.docstatus    = 1
            AND ss.start_date = %(start_date)s
            {conditions}
        ORDER BY
            ss.employee_name
    """.format(conditions=conditions), query_params, as_dict=1)

    if not rows:
        return []

    # --- Enrich with Employee master data ---
    employee_ids = list({r.employee_id for r in rows})

    cl_records = frappe.db.get_all(
        "Company Link",
        filters={"name": ["in", employee_ids]},
        fields=["name", "employee"]
    )
    cl_map = {r.name: r.employee for r in cl_records}

    emp_master_ids = [v for v in cl_map.values() if v]
    emp_details_map = {}
    if emp_master_ids:
        emp_records = frappe.db.get_all(
            "Employee",
            filters={"name": ["in", emp_master_ids]},
            fields=["name", "cell_number", "aadhar_number"]
        )
        emp_details_map = {r.name: r for r in emp_records}

    data             = []
    total_net_salary = 0.0
    total_lwf        = 0.0

    for idx, row in enumerate(rows, start=1):
        emp_master_id = cl_map.get(row.employee_id)
        emp_detail    = emp_details_map.get(emp_master_id, {})

        net = flt(row.net_salary, 2)
        lwf = flt(row.lwf_amount, 2)

        total_net_salary += net
        total_lwf        += lwf

        data.append({
            "sr_no":         idx,
            "employee_id":   row.employee_id,
            "employee_name": row.employee_name,
            "mobile_no":     emp_detail.get("cell_number") or "-",
            "aadhar_no":     emp_detail.get("aadhar_number") or "-",
            "net_salary":    net,
            "lwf_amount":    lwf,
        })

    # Totals row
    data.append({
        "sr_no":         "",
        "employee_id":   "",
        "employee_name": "Total",
        "mobile_no":     "",
        "aadhar_no":     "",
        "net_salary":    flt(total_net_salary, 2),
        "lwf_amount":    flt(total_lwf, 2),
        "bold":          1,
    })

    return data


@frappe.whitelist()
def get_lwf_employees_for_filter(year=None, month=None, companies=None, txt=""):
    """
    Called by the Employee MultiSelectList get_data in JS.
    Returns employees who have a submitted slip with LWF > 0 for the period.
    """
    if not month or not year:
        return []

    month_num = MONTH_MAP.get(month)
    if not month_num:
        return []

    import json
    if isinstance(companies, str):
        try:
            companies = json.loads(companies)
        except Exception:
            companies = []

    start_date = f"{year}-{month_num:02d}-01"
    txt_filter = f"%{txt}%"

    company_condition = ""
    params = {
        "component":  COMPONENT_NAME,
        "start_date": start_date,
        "txt":        txt_filter,
    }

    if companies:
        company_condition = "AND ss.company IN %(companies)s"
        params["companies"] = tuple(companies)

    results = frappe.db.sql("""
        SELECT DISTINCT
            ss.employee,
            ss.employee_name
        FROM
            `tabSalary Slip` ss
        INNER JOIN
            `tabSalary Details` sd
            ON  sd.parent           = ss.name
            AND sd.parenttype       = 'Salary Slip'
            AND sd.parentfield      = 'deductions'
            AND sd.salary_component = %(component)s
            AND sd.amount > 0
        WHERE
            ss.docstatus    = 1
            AND ss.start_date = %(start_date)s
            AND (ss.employee LIKE %(txt)s OR ss.employee_name LIKE %(txt)s)
            {company_condition}
        ORDER BY ss.employee_name
        LIMIT 50
    """.format(company_condition=company_condition), params, as_dict=1)

    return results

File: saral_hr/saral_hr/report/labour_welfare_fund_register/labour_welfare_fund_register.js
// Copyright (c) 2026, sj and contributors
// For license information, please see license.txt

frappe.query_reports["Labour Welfare Fund Register"] = {
    filters: [
        {
            fieldname: "year",
            label: __("Year"),
            fieldtype: "Select",
            options: get_year_options(),
            reqd: 1
        },
        {
            fieldname: "month",
            label: __("Month"),
            fieldtype: "Select",
            options: [
                "", "January", "February", "March", "April",
                "May", "June", "July", "August",
                "September", "October", "November", "December"
            ],
            reqd: 1
        },
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "MultiSelectList",
            get_data: function(txt) {
                return frappe.db.get_link_options("Company", txt);
            }
        },
        {
            fieldname: "employee",
            label: __("Employee"),
            fieldtype: "MultiSelectList",
            get_data: function(txt) {
                let year      = frappe.query_report.get_filter_value("year");
                let month     = frappe.query_report.get_filter_value("month");
                let companies = frappe.query_report.get_filter_value("company");

                return frappe.call({
                    method: "saral_hr.saral_hr.report.labour_welfare_fund_register.labour_welfare_fund_register.get_lwf_employees_for_filter",
                    args: {
                        year:      year,
                        month:     month,
                        companies: JSON.stringify(companies || []),
                        txt:       txt || ""
                    }
                }).then(r => {
                    return (r.message || []).map(emp => ({
                        value:       emp.employee,
                        description: emp.employee_name || ""
                    }));
                });
            }
        }
    ],

    // ── Custom print ──────────────────────────────────────────────────────────
    onload: function(report) {
        // Wait for the toolbar to render then replace the Print button action
        frappe.after_ajax(() => {
            report.page.set_primary_action(__("Print"), () => {
                lwf_custom_print(report);
            }, "printer");
        });
    }
};

// ─────────────────────────────────────────────────────────────────────────────

function get_year_options() {
    let current_year = new Date().getFullYear();
    let years = [""];
    for (let i = current_year - 2; i <= current_year + 1; i++) {
        years.push(i.toString());
    }
    return years;
}

function lwf_custom_print(report) {
    let data    = frappe.query_report.data || [];
    let filters = frappe.query_report.get_values() || {};

    let month   = filters.month || "";
    let year    = filters.year  || "";

    // Separate data rows from the totals row
    let rows       = data.filter(r => r.employee_name !== "Total");
    let totals_row = data.find(r => r.employee_name === "Total") || {};

    // ── Build table rows ──
    let tbody = rows.map((row, idx) => `
        <tr>
            <td class="center">${idx + 1}</td>
            <td>${row.employee_id || ""}</td>
            <td>${row.employee_name || ""}</td>
            <td class="center">${row.mobile_no || "-"}</td>
            <td class="center">${row.aadhar_no || "-"}</td>
            <td class="num">${fmt(row.net_salary)}</td>
            <td class="num">${fmt(row.lwf_amount)}</td>
        </tr>
    `).join("");

    // ── Totals row ──
    tbody += `
        <tr class="total-row">
            <td colspan="5" style="text-align:right; padding-right:12px; font-weight:bold;">Total</td>
            <td class="num">${fmt(totals_row.net_salary)}</td>
            <td class="num">${fmt(totals_row.lwf_amount)}</td>
        </tr>
    `;

    // ── Full HTML ──
    let html = `
        <!DOCTYPE html>
        <html>
        <head>
            <meta charset="UTF-8">
            <title>Labour Welfare Fund Register – ${month} ${year}</title>
            <style>
                * { margin:0; padding:0; box-sizing:border-box; }
                body { font-family: Arial, sans-serif; font-size: 11px; color: #000; }
                .container { padding: 24px; }

                .header { text-align: center; margin-bottom: 16px; }
                .header .company-name { font-size: 16px; font-weight: bold; text-transform: uppercase; }
                .header .report-title { font-size: 13px; font-weight: bold; margin-top: 4px; }
                .header .period       { font-size: 11px; margin-top: 2px; color: #444; }

                table { width: 100%; border-collapse: collapse; margin-top: 12px; }

                thead tr th {
                    background-color: #e8e8e8;
                    border: 1px solid #999;
                    padding: 7px 8px;
                    text-align: center;
                    font-size: 11px;
                    font-weight: bold;
                }

                tbody tr td {
                    border: 1px solid #bbb;
                    padding: 6px 8px;
                    font-size: 11px;
                }

                tbody tr:nth-child(even):not(.total-row) { background-color: #f9f9f9; }

                td.num    { text-align: right; }
                td.center { text-align: center; }

                tr.total-row td {
                    font-weight: bold;
                    background-color: #eef4fb;
                    border-top: 2px solid #555;
                    border-bottom: 2px solid #555;
                }

                .footer {
                    margin-top: 40px;
                    display: flex;
                    justify-content: space-between;
                }
                .sign-block { text-align: center; width: 180px; }
                .sign-block .line  { border-top: 1px solid #000; margin-bottom: 4px; }
                .sign-block .label { font-size: 10px; color: #444; }

                @media print {
                    @page { size: A4 landscape; margin: 15mm; }
                    body  { -webkit-print-color-adjust: exact; print-color-adjust: exact; }
                }
            </style>
        </head>
        <body>
        <div class="container">

            <div class="header">
                <div class="company-name">${frappe.boot.sysdefaults.company || ""}</div>
                <div class="report-title">Labour Welfare Fund Register</div>
                <div class="period">For the Month of ${month} ${year}</div>
            </div>

            <table>
                <thead>
                    <tr>
                        <th style="width:55px;">Sr. No.</th>
                        <th style="width:140px;">Employee ID</th>
                        <th style="width:200px;">Employee Name</th>
                        <th style="width:130px;">Mobile No</th>
                        <th style="width:160px;">Aadhar No</th>
                        <th style="width:130px;">Net Salary</th>
                        <th style="width:120px;">LWF Amount</th>
                    </tr>
                </thead>
                <tbody>
                    ${tbody}
                </tbody>
            </table>

            <div class="footer">
                <div class="sign-block">
                    <div class="line"></div>
                    <div class="label">Prepared By</div>
                </div>
                <div class="sign-block">
                    <div class="line"></div>
                    <div class="label">Checked By</div>
                </div>
                <div class="sign-block">
                    <div class="line"></div>
                    <div class="label">Authorised Signatory</div>
                </div>
            </div>

        </div>
        </body>
        </html>
    `;

    // ── Open in new tab and trigger browser print ──
    let w = window.open("", "_blank");
    w.document.write(html);
    w.document.close();
    w.focus();
    setTimeout(() => { w.print(); }, 500);
}

function fmt(val) {
    return parseFloat(val || 0).toFixed(2);
}